complete.py

```diff
@@ -0,0 +1,795 @@
+        self.particle_state_subset.apply_to_context(self.context_subset, ignore_velocities=True) #apply the subset particle state to its context
+        self.particle_state_subset.update_from_context(self.context_subset, ignore_velocities=True) #update the subset particle state from its context to updated the potential energy
+
+        #get the reduced potential
+        reduced_potential = self._compute_hybrid_potential(_lambda = self._iteration / self._n_iterations, particle_state = particle_state)
+        perturbed_reduced_potential = self._compute_hybrid_potential(_lambda = (self._iteration + 1.0) / self._n_iterations, particle_state = particle_state)
+        self._current_state_works.append(self._current_state_works[-1] + (perturbed_reduced_potential - reduced_potential))
+
+        #make a new force object
+        mm_force_matrix = self._compute_hybrid_forces(_lambda = (self._iteration + 1.0) / self._n_iterations, particle_state = particle_state).value_in_unit_system(unit.md_unit_system)
+        self.integrator.setPerDofVariableByName('modified_force', mm_force_matrix)
+
+        #report
+        if self._write_trajectory:
+            self.particle.update_state(particle_state)
+            self.reporter.record([self.particle])
+
+        #logger
+        #self._log_context_parameters()
+
+
+
+    def _during_integration(self, *args, **kwargs):
+        particle_state = args[0]
+        self._iteration += 1.0
+
+
+        #update the particle state and the particle state subset
+        particle_state.update_from_context(self.context, ignore_velocities=True) #update the particle state from the context
+        self.particle_state_subset.positions = particle_state.positions[list(self._subset_indices_map.keys())] #update the particle subset positions appropriately
+        self.particle_state_subset.apply_to_context(self.context_subset, ignore_velocities=True) #apply the subset particle state to its context
+        self.particle_state_subset.update_from_context(self.context_subset, ignore_velocities=True) #update the subset particle state from its context to updated the potential energy
+
+        #get the reduced potential
+        if self._iteration < self._n_iterations:
+            reduced_potential = self._compute_hybrid_potential(_lambda = self._iteration / self._n_iterations, particle_state = particle_state)
+            perturbed_reduced_potential = self._compute_hybrid_potential(_lambda = (self._iteration + 1.0) / self._n_iterations, particle_state = particle_state)
+            self._current_state_works.append(self._current_state_works[-1] + (perturbed_reduced_potential - reduced_potential))
+
+            #and create a new modified force
+            mm_force_matrix = self._compute_hybrid_forces(_lambda = (self._iteration + 1.0) / self._n_iterations, particle_state = particle_state).value_in_unit_system(unit.md_unit_system)
+            self.integrator.setPerDofVariableByName('modified_force', mm_force_matrix)
```

Comment:
I am not sure what happens here -- could you explain this call a bit?

Reply:
since we don't use the context's `f` variable in the `V` step of the integration scheme, we have to create a `modified_f` that is a `perDofVariable` in the integrator which contains the force vector of the mm systems _augmented_ by the force matrix of the ani system. `_compute_hybrid_forces` creates this hybrid force matrix given the positions of the full system, and the value of lambda

Reply:
then the modified force is passed to the integrator in the `setPerDofVariableByName`

Reply:
thanks for clarifying! My confusion came from the setPerDofVariableByName; I was simply not sure why this gets the modified force, but now that makes sense!